core/src/main/java/tc/oc/pgm/api/filter/Filter.java

```diff
@@ -1,14 +1,53 @@
 package tc.oc.pgm.api.filter;
 
+import com.google.common.collect.ImmutableList;
+import java.util.Collection;
+import org.bukkit.event.Event;
+import tc.oc.pgm.api.feature.Feature;
+import tc.oc.pgm.api.feature.FeatureReference;
 import tc.oc.pgm.api.filter.query.Query;
+import tc.oc.pgm.filters.FilterParser;
 
+/**
+ * Something that can answer "yes", "no" or "i don't care" to the imagined question: "Can X
+ * happen?". To be able to answer the question context is provided initially through a {@link Query}
+ * and then processed by the filter itself before a response is given.
+ *
+ * <p>If a filter is connected to a specific {@link Feature} a {@link FeatureReference} should be
+ * passed at xml parse time (when the filter is created) and {@link FeatureReference#get() fetched}
+ * at match time
+ *
+ * <p>For examples on how to create filters see {@link FilterParser}
+ *
+ * @see Query
+ */
 public interface Filter {
 
+  /**
+   * Filters with children are responsible for returning the events of their children.
+   *
+   * <p>Empty list tells us that the filter is not dynamic
+   */
+  default Collection<Class<? extends Event>> getRelevantEvents() {
+    return ImmutableList.of();
+  }
+
   /** Least-derived query type that this filter might not abstain from */
   Class<? extends Query> getQueryType();
 
   QueryResponse query(Query query);
 
+  default boolean response(Query query) {
+    switch (query(query)) {
+      case ALLOW:
+        return true;
+      case DENY:
+        return false;
+      default:
+        throw new UnsupportedOperationException("Filter did not respond to the query");
```

Comment:
Is there a reason we need to make this throw? (having functions that returns booleans that throw are not nice)

Could this just be `isAllowed` ?

Reply:
https://github.com/OvercastNetwork/ProjectAres/blob/master/PGM/src/main/java/tc/oc/pgm/filters/Filter.java#L29

This was taken from 1.9+ PGM, i think it works as intended, you should **only** use this shortcut method if you know it **will** give a response, `assertRespondsTo(Class)` method is mentioned in the original but that method is not here.

Also keep in mind for all of the dynamic filter stuff response **has** to be either allow or deny, if it's neither when do you tell it's a raise and when it's a fall? It's essentially an exception for unsupported case, which is justified, i think the javadoc should make it clear that this is for special cases only where you know filter won't abstain, and to use #query instead for general use.

Regarding using just `isAllowed`, that would be hiding bugs under the carpet, if a filter abstains it would be treated as a deny, triggering a fall, and when back to allow it would trigger a raise

Reply:
Sounds good.